### PR TITLE
Configured golangci-lint

### PR DESCRIPTION
**Pull request by Opelord**
Issue: KNTK-192

### CODE REVIEW

**Comment by Fenthick** [comment]
Please squash the commits

**Comment by Fenthick** [comment]
@Opelord Please rename the commit to e.g. _Configure golangci-lint and adjust the code to comply_

Also, take a look: https://chris.beams.io/posts/git-commit/

File: apiv6/localhost_apiserver/main.go
Comment (mmac-m3a) [inline apiv6/localhost_apiserver/main.go:41]: It is interesting that the code worked before without any value being returned  ;-).

**Review by mmac-m3a — APPROVED** [review]
I hope those linters are worth those 101 mouse clicks :-).

**Comment by mmac-m3a** [comment]
Waiting for @Fenthick to approve before merging.

File: apiv6/localhost_apiserver/main.go
Comment (Fenthick) [reply apiv6/localhost_apiserver/main.go:41]: It was a naked return: https://tour.golang.org/basics/7 - possible, but unreadable

File: apiv6/kentikapi/httputil/httputil_test.go
Comment (Fenthick) [inline apiv6/kentikapi/httputil/httputil_test.go:39]: Please remove that. The error is already checked in //assert section.

File: apiv6/kentikapi/httputil/httputil_test.go
Comment (Fenthick) [inline apiv6/kentikapi/httputil/httputil_test.go:26]:
```suggestion
```
Non-nil error is valuable information itself

File: apiv6/kentikapi/httputil/httputil_test.go
Comment (Fenthick) [inline apiv6/kentikapi/httputil/httputil_test.go:110]: Ditto

File: apiv6/kentikapi/httputil/httputil_test.go
Comment (Fenthick) [inline apiv6/kentikapi/httputil/httputil_test.go:51]: Ditto

**Review by Fenthick — CHANGES_REQUESTED** [review]
Partially reviewed.

Renaming package names leads to too big diff - it is too hard to review. Please revert package renames (and suppress linter).

File: apiv6/kentikapi/httputil/httputil_test.go
Comment (Fenthick) [reply apiv6/kentikapi/httputil/httputil_test.go:39]: Moreover this breaks the test - it does not check anything now

File: apiv6/examples/synthetics/mesh_test/metrics_matrix.go
Comment (Fenthick) [inline apiv6/examples/synthetics/mesh_test/metrics_matrix.go:7]:
```suggestion
// metricsMatrix holds "fromAgent" -> "toAgent" connection metrics
```

File: apiv6/kentikapi/client.go
Comment (Fenthick) [inline apiv6/kentikapi/client.go:37]:
```suggestion
	// LogPayloads enables logging of request and response payloads.
```

File: apiv6/localhost_apiserver/cloudexport/go/routers.go
Comment (Fenthick) [inline apiv6/localhost_apiserver/cloudexport/go/routers.go:33]:
```suggestion
// Routes are a collection of defined api endpoints
```

File: apiv6/localhost_apiserver/synthetics/go/routers.go
Comment (Fenthick) [inline apiv6/localhost_apiserver/synthetics/go/routers.go:33]:
```suggestion
// Routes are a collection of defined api endpoints
```

File: apiv6/kentikapi/client_integration_test.go
Comment (Opelord) [reply apiv6/kentikapi/client_integration_test.go:28]: Like this?

```
//nolint:errcheck // Conflict with defer Close() (additional info: https://github.com/kisielk/errcheck/issues/55)
defer httpResp.Body.Close()
```

**Review by Fenthick — CHANGES_REQUESTED** [review]

File: .golangci.yml
Comment (Fenthick) [reply .golangci.yml]: Done

File: .golangci.yml
Comment (Fenthick) [reply .golangci.yml]: Done

File: .golangci.yml
Comment (Fenthick) [reply .golangci.yml]: Done

File: .golangci.yml
Comment (Fenthick) [reply .golangci.yml]: Done

File: .golangci.yml
Comment (Fenthick) [reply .golangci.yml]: Done

File: .golangci.yml
Comment (Fenthick) [reply .golangci.yml]: Done

File: .golangci.yml
Comment (Fenthick) [reply .golangci.yml]: Done

File: .golangci.yml
Comment (Fenthick) [reply .golangci.yml]: Done

File: .golangci.yml
Comment (Fenthick) [reply .golangci.yml]: Done

File: .golangci.yml
Comment (Fenthick) [reply .golangci.yml]: Done

File: apiv6/kentikapi/client.go
Comment (Fenthick) [reply apiv6/kentikapi/client.go]: Done

File: apiv6/kentikapi/client_integration_test.go
Comment (Fenthick) [reply apiv6/kentikapi/client_integration_test.go]: Done

File: apiv6/kentikapi/client_integration_test.go
Comment (Fenthick) [reply apiv6/kentikapi/client_integration_test.go]: Yes, done.

File: apiv6/kentikapi/httputil/httputil_test.go
Comment (Fenthick) [reply apiv6/kentikapi/httputil/httputil_test.go]: Done

File: apiv6/kentikapi/httputil/httputil_test.go
Comment (Fenthick) [reply apiv6/kentikapi/httputil/httputil_test.go]: Done

File: apiv6/kentikapi/httputil/httputil_test.go
Comment (Fenthick) [reply apiv6/kentikapi/httputil/httputil_test.go]: Done

File: apiv6/kentikapi/httputil/httputil_test.go
Comment (Fenthick) [reply apiv6/kentikapi/httputil/httputil_test.go]: Done

File: kentikapi/models/device.go
Comment (Fenthick) [inline kentikapi/models/device.go:15]: Ditto: put comment above

File: kentikapi/tags_integration_test.go
Comment (Fenthick) [inline kentikapi/tags_integration_test.go:1]: Please put it next to duplicated code

File: kentikapi/users_integration_test.go
Comment (Fenthick) [inline kentikapi/users_integration_test.go:1]: Please put it next to duplicated code

**Review by Fenthick — CHANGES_REQUESTED** [review]
Finished

File: kentikapi/internal/resources/custom_applications.go
Comment (Opelord) [reply kentikapi/internal/resources/custom_applications.go:48]: This nolint is addressing the _if_ statement below, not the function declaration.

Comment (Opelord) [reply kentikapi/internal/resources/custom_applications.go:48]: I moved it accordingly above this line to avoid confusion later

Comment (Opelord) [reply kentikapi/internal/resources/custom_applications.go:48]: Nevermind, I didn't know you could linebreak ifs like in the comment below.

File: kentikapi/internal/api_payloads/query.go
Comment (mateuszmidor) [reply kentikapi/internal/api_payloads/query.go:150]: Error should be passed up the call chain

Comment (Fenthick) [reply kentikapi/internal/api_payloads/query.go:150]: @Opelord Could you fix that here? It should be straightforward to pass the error up the call stack.

Comment (Opelord) [reply kentikapi/internal/api_payloads/query.go:150]: Yes, I will.

File: .golangci.yml
Comment (Fenthick) [inline .golangci.yml:26]: I wasn't aware that those files are duplicated as a whole. In that case whole file could be excluded.  I checked in docs that there are no package-level excludes. So the simplest solution is to put nolint statement in a file (but maybe not above `package` statement so that it would not be a package docstring, if possible).

https://golangci-lint.run/usage/false-positives/

**Review by Fenthick — CHANGES_REQUESTED** [review]

File: apiv6/examples/synthetics/mesh_test/metrics_matrix.go
Comment (Fenthick) [reply apiv6/examples/synthetics/mesh_test/metrics_matrix.go]: Done

File: apiv6/kentikapi/client.go
Comment (Fenthick) [reply apiv6/kentikapi/client.go]: Done

File: apiv6/localhost_apiserver/cloudexport/go/routers.go
Comment (Fenthick) [reply apiv6/localhost_apiserver/cloudexport/go/routers.go]: Done

File: apiv6/localhost_apiserver/synthetics/go/routers.go
Comment (Fenthick) [reply apiv6/localhost_apiserver/synthetics/go/routers.go]: Done

File: kentikapi/internal/api_payloads/interface.go
Comment (Fenthick) [reply kentikapi/internal/api_payloads/interface.go]: Done

File: kentikapi/internal/api_payloads/query.go
Comment (Fenthick) [reply kentikapi/internal/api_payloads/query.go]: Done

File: kentikapi/internal/api_payloads/saved_filter.go
Comment (Fenthick) [reply kentikapi/internal/api_payloads/saved_filter.go]: Done

File: kentikapi/internal/resources/alerting.go
Comment (Fenthick) [reply kentikapi/internal/resources/alerting.go]: Done

File: kentikapi/internal/resources/custom_applications.go
Comment (Fenthick) [reply kentikapi/internal/resources/custom_applications.go]: Done

File: kentikapi/internal/resources/custom_applications.go
Comment (Fenthick) [reply kentikapi/internal/resources/custom_applications.go]: Done

File: kentikapi/internal/resources/custom_applications.go
Comment (Fenthick) [reply kentikapi/internal/resources/custom_applications.go]: Done

File: kentikapi/internal/resources/custom_dimensions.go
Comment (Fenthick) [reply kentikapi/internal/resources/custom_dimensions.go]: Done

File: kentikapi/internal/resources/devices_test.go
Comment (Fenthick) [reply kentikapi/internal/resources/devices_test.go]: Done

File: kentikapi/internal/resources/my_kentik_portal.go
Comment (Fenthick) [reply kentikapi/internal/resources/my_kentik_portal.go]: Done

File: kentikapi/internal/resources/tags.go
Comment (Fenthick) [reply kentikapi/internal/resources/tags.go]: Done

File: kentikapi/internal/resources/users.go
Comment (Fenthick) [reply kentikapi/internal/resources/users.go]: Done

File: kentikapi/models/device.go
Comment (Fenthick) [reply kentikapi/models/device.go]: Done

File: kentikapi/tags_integration_test.go
Comment (Fenthick) [reply kentikapi/tags_integration_test.go]: Done

File: kentikapi/users_integration_test.go
Comment (Fenthick) [reply kentikapi/users_integration_test.go]: Done

File: .golangci.yml
Comment (Opelord) [reply .golangci.yml:26]: It seems //nolint directive is [ignored by godoc](https://github.com/golangci/golangci-lint/issues/892).
I will revert the changes to the previous commit as there is also no other way to nolint an entire file in another place.

File: kentikapi/internal/api_payloads/query.go
Comment (Fenthick) [reply kentikapi/internal/api_payloads/query.go]: Done

File: .golangci.yml
Comment (Fenthick) [reply .golangci.yml]: Done

**Review by Fenthick — APPROVED** [review]

**Comment by Opelord** [comment]
@mmac-m3a Daniel approved the changes. Can you approve too, because your review got cancelled after the first push with corrections.

File: kentikapi/internal/api_payloads/saved_filter.go
Comment (mmac-m3a) [inline kentikapi/internal/api_payloads/saved_filter.go:80]: The logic behind those renamings escapes me. At one place we are changing from (ugly) `camelCase` to (little bit less ugly) `CamelCase` (e.g. [here](https://github.com/kentik/community_sdk_golang/pull/51/files#diff-65234495dfcbcdeb9177db2e28015c5a28515803ef483902a565847f1cbc1ceeR21))... and here it is the other way around. Is this forced by the linters (aren't they fighting each other)? Is this a type-name vs. function-name thing?

File: kentikapi/internal/resources/tags.go
Comment (mmac-m3a) [reply kentikapi/internal/resources/tags.go:1]: Where was this fixed? I'm, not seeing it.

File: kentikapi/internal/resources/users.go
Comment (mmac-m3a) [reply kentikapi/internal/resources/users.go:1]: Same here

**Review by mmac-m3a — APPROVED** [review]
I cannot look at it anymore ... and definitely don't want to go though 80+ clicks again :-).
... but, please, answer my questions.

File: kentikapi/internal/api_payloads/saved_filter.go
Comment (Opelord) [reply kentikapi/internal/api_payloads/saved_filter.go:80]: In Golang the difference between `CamelCase` and `camelCase` is the difference between an `exported` and `unexported` field, struct, function. It is genereally discouraged to have an exported function (a function that can be used by other packages) return an unexported struct. 
Now in this case we had an exported function return an unexported type but the function wasn't used in other packages. So instead of exporting the return I "unexported" (if this is even a word) the function.
In case of RestClient we came to conclusion that because functions returning this struct are used in other packages than the return type should be exported too.
https://github.com/golang/lint/issues/210

**Comment by Opelord** [comment]
As for your question [here](https://github.com/kentik/community_sdk_golang/pull/51/#discussion_r707817738) it was solved in [this conversation.](https://github.com/kentik/community_sdk_golang/pull/51/files/6f8129d4106cf250058c83afd6fdd215d9948d9e#r706051441) The problem was that I was supposed to avoid package level nolints. The logic being that it's better to nolint only the block of code that is called by the linter. However in these files the linter highlighted all of their content as a duplicate of the other. I tried to move the exception to the linters configuration file but it hurt clarity of what is actually happening so we decided to move it back. Hence probably the confusion that nothing changed.

**Pull request merged**
